### ISSUE TITLE
assign_op_pattern false positive for Assign trait impl

### ISSUE THREAD

**Issue by crumblingstatue** [Manishearth/rust-clippy#1302]
We start out with

``` rust
use std::ops::{Mul, MulAssign};

#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Wrap(i64);

impl Mul<i64> for Wrap {
    type Output = Self;

    fn mul(self, rhs: i64) -> Self {
        Wrap(self.0 * rhs)
    }
}

impl MulAssign<i64> for Wrap {
    fn mul_assign(&mut self, rhs: i64) {
        *self = *self * rhs
    }
}

#[test]
fn test() {
    let mut foo = Wrap(5);
    foo *= 10;
    assert_eq!(foo, Wrap(50));
}

```

Clippy result:

```
warning: manual implementation of an assign operation, #[warn(assign_op_pattern)] on by default
  --> src/lib.rs:16:9
   |
16 |         *self = *self * rhs
   |         ^^^^^^^^^^^^^^^^^^^
   |
help: replace it with
   |         *self *= rhs

```

Running test after applying the suggestion:

```
warning: function cannot return without recurring, #[warn(unconditional_recursion)] on by default
  --> src/lib.rs:15:5
   |
15 |     fn mul_assign(&mut self, rhs: i64) {
   |     ^
   |
note: recursive call site
  --> src/lib.rs:16:9
   |
16 |         *self *= rhs
   |         ^^^^^^^^^^^^
   = help: a `loop` may express intention better if this is on purpose

    Finished debug [unoptimized + debuginfo] target(s) in 0.76 secs
     Running target/debug/deps/assignop-22d33742ac4da442

running 1 test

thread 'test' has overflowed its stack
fatal runtime error: stack overflow

```

**Comment by llogiq** [comment]
Could `self.0 *= rhs` work?

**Comment by crumblingstatue** [comment]
> Could self.0 *= rhs work?

The lint doesn't only trigger for newtype structs.

Theoretically, you could have a type that doesn't have any underlying type that implements `Mul`, such as:
```rust
/// Some kind of super ultra in-ram arbitrary precision numeric type
#[derive(Copy, Clone, Debug, PartialEq)]
pub struct SuperNum;

impl Mul<i64> for SuperNum {
    type Output = Self;

    fn mul(self, _rhs: i64) -> Self {
        // Who knows how such a type is implemented
        unimplemented!()
    }
}
```

**Comment by llogiq** [comment]
Isn't there some sort of blanket impl for `Mul` types that does the same?
